### ISSUE TITLE
SIGABRT from DisplayLinkCallback (SDDisplayLink.m:255:27) because CVDisplayLinkRef has been deallocated

### ISSUE THREAD

**Issue by renewedvision-kh** [SDWebImage/SDWebImage#3530]
### New Issue Checklist

* [ ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | macos
 Platform Version        | 13.3.1
 SDWebImage Version      | 5.15.7
 Integration Method      | SPM
 Xcode Version           | Xcode 14.3
 Repro rate              | timing related
 Repro with our demo prj | timing related
 Demo project link       | none

### Issue Description and Steps

`SDAnimatedImageView.setImage` will set `player` to `nil` if a player is no longer needed (approximately line 124):
```
    if (!self.isProgressive) {
        // Stop animating
        self.player = nil;
        self.currentFrame = nil;
        self.currentFrameIndex = 0;
        self.currentLoopCount = 0;
    }
```
`SDAnimatedImagePlayer.dealloc` does nothing special if its `displayLink` property has been set so it will simply be deallocated. `SDDisplayLink.dealloc` will simply call `CVDisplayLinkRelease` releasing the `CVDisplayLinkRef`; however, no code has called `CVDisplayLinkStop` to actually stop callbacks to `SDDisplayLink.DisplayLinkCallback`.

It is unclear if the `SDAnimatedImageView` needs to call its `stopAnimating` before setting `player` to `nil`, if `SDAnimatedImagePlayer` needs to call `stopPlaying` in its `dealloc` or if `SDDisplayLink` needs to call its `stop` before calling `CVDisplayLinkRelease` in its `dealloc`. My personal feeling is that `SDDisplayLink` should be responsible for not leaving itself in a state that might crash.

The scenario where this occurs is I have an `SDAnimatedImageView` that animates an image the mouse hovers over the view and shows a different still image when the mouse is not over the view. When the mouse enters the view the following code (Swift) sets the animated image:
```
animatedImageView.sd_setImage(with: url, placeholderImage: placeholderImage, options: .progressiveLoad)
```
When the mouse exits the view the following code (Swift) sets the still image:
```
animatedImageView.sd_setImage(with: asset?.thumbnailURL)
```
I both cases the URL references a WebP image the first is a progressive image while the second is a still image; however, I do not believe it matters what encoding is used just that the first is an animated image and the second is not. The crash randomly (timing) occurs when the mouse leaves and the second `sd_setImage` is called. The conditions are as described above where `SDAnimatedImageView.setImage` will dealloc its `player.

**Comment by dreampiggy** [comment]
@renewedvision-kh  Can you test that whether the #3524 branch fix the issue ?

**Comment by dreampiggy** [comment]
I think this is macOS only issue.

The CVDisplayLink is C level API, it need to handle the case when the display link itself dealloc during the callback function `displayLinkDidRefresh`

**Comment by dreampiggy** [comment]
I think the #3531 will indeed fix your issue. This is a bug because of wrong handle for C callback, should use the similiar solution like iOS/watchOS using weak proxy

**Comment by dreampiggy** [comment]
Available in 5.15.8

**Comment by renewedvision-kh** [comment]
Awesome, the issue is fixed for me. I see `-[SDDisplayLink dealloc]` called and it now calls `CVDisplayLinkStop(_displayLink);` to stop the callbacks. No Crash! Thank you @dreampiggy.
